iree/hal/driver_registry.c

```diff
@@ -60,6 +60,39 @@ iree_hal_driver_registry_default() {
   return &iree_hal_driver_registry_default_;
 }
 
+// Creates and initializes an empty device driver registry
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_driver_registry_create(
+    iree_allocator_t allocator, iree_hal_driver_registry_t** out_registry) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(
+      allocator, sizeof(struct iree_hal_driver_registry_s),
+      (void**)out_registry));
+  iree_slim_mutex_initialize(&((*out_registry)->mutex));
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_driver_registry_unregister_and_free(
+    iree_hal_driver_registry_t* registry, iree_allocator_t allocator) {
+  IREE_ASSERT_ARGUMENT(registry);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // unregister each factory
+  for (iree_host_size_t i = 0; i < registry->factory_count; ++i) {
```

Comment:
this loop won't work as-is because the unregister mutates the factory_count and the list - you will need to do this backwards (count -> 0) so you are unregistering the tail each iteration or change it to a while-loop while the factory_count > 0